driver-core/src/main/com/mongodb/internal/connection/ScramShaAuthenticator.java

```diff
@@ -246,25 +254,15 @@ private byte[] h(final byte[] data) throws SaslException {
             }
         }
 
-        private byte[] hi(final byte[] password, final byte[] salt, final int iterations) throws SaslException {
+        private byte[] hi(final String password, final byte[] salt, final int iterations) throws SaslException {
             try {
-                SecretKeySpec key = new SecretKeySpec(password, hmacAlgorithm);
-                Mac mac = Mac.getInstance(hmacAlgorithm);
-                mac.init(key);
-                mac.update(salt);
-                mac.update(INT_1);
-                byte[] result = mac.doFinal();
-                byte[] previous = null;
-                for (int i = 1; i < iterations; i++) {
-                    mac.update(previous != null ? previous : result);
-                    previous = mac.doFinal();
-                    xorInPlace(result, previous);
-                }
-                return result;
+                SecretKeyFactory secretKeyFactory = SecretKeyFactory.getInstance(pbeAlgorithm);
+                PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, iterations, keyLength);
+                return secretKeyFactory.generateSecret(spec).getEncoded();
             } catch (NoSuchAlgorithmException e) {
-                throw new SaslException(format("Algorithm for '%s' could not be found.", hmacAlgorithm), e);
-            } catch (InvalidKeyException e) {
-                throw new SaslException(format("Invalid key for %s", hmacAlgorithm), e);
+                throw new SaslException(format("Algorithm for '%s' could not be found.", pbeAlgorithm), e);
```

Comment:
[nit] This String is repeated in the code. It could be a candidate for a constant.